init.go

```diff
@@ -200,3 +209,29 @@ func createDirectories(directories []string) error {
 	}
 	return nil
 }
+
+func shutdownPlugin() {
+	log.Info("Shutting down apidAnalytics plugin")
+
+	// Close all open files and move directories in tmp to staging.
+	bucketMaplock.RLock()
+	defer bucketMaplock.RUnlock()
+	for _, bucket := range bucketMap {
+		log.Infof("closing bucket '%s' as a part of shutdown", bucket.DirName)
+		closeGzipFile(bucket.FileWriter)
+
+		dirToBeClosed := filepath.Join(localAnalyticsTempDir, bucket.DirName)
+		stagingPath := filepath.Join(localAnalyticsStagingDir, bucket.DirName)
+		// close files in tmp folder and move directory to
+		// staging to indicate its ready for upload
+		err := os.Rename(dirToBeClosed, stagingPath)
+		if err != nil {
+			log.Errorf("Cannot move directory '%s' from"+
+				" tmp to staging folder due to '%s", bucket.DirName, err)
+		} else {
+			bucketMaplock.Lock()
```

Comment:
You can't `Lock()` before you `RUnlock()`, it's a deadlock.
Since this is shutdown, I would suggest not delete the keys one-by-one. Just reset the map at the end of this func, after the `for` loop:
```
bucketMaplock.Lock()
bucketMap = nil
bucketMaplock.Unlock()
```

Be sure to `RUnlock` before you do this.